### PR TITLE
Refactor of the DER buffer handling

### PR DESCRIPTION
**Pull request by dgarske**
Added new DerBuffer struct that includes the type and heap ptr. Added new InitDer, AllocDer and FreeDer functions. Cleanup of some missing "heap" args on XMALLOC/XFREE. In FreeDer uses ForceZero if type is private key.

### CODE REVIEW

**Comment by toddouska** [comment]
Thanks, will review within the next couple days.

File: src/internal.c [hunk 1]
```diff
@@ -4493,14 +4469,8 @@ int CopyDecodedToX509(WOLFSSL_X509* x509, DecodedCert* dCert)
     }
 
     /* store cert for potential retrieval */
-    x509->derCert.buffer = (byte*)XMALLOC(dCert->maxIdx, NULL,
-                                          DYNAMIC_TYPE_CERT);
-    if (x509->derCert.buffer == NULL) {
-        ret = MEMORY_E;
-    }
-    else {
+    if (AllocDer(&x509->derCert, dCert->maxIdx, CERT_TYPE, NULL) == 0) {
```
Comment (toddouska) [inline src/internal.c]: Doesn't look like MEMORY_E is being set/returned on failure like it used to.  Is this intended?

**Comment by toddouska** [comment]
Looks good, just the one question/comment above.

File: src/internal.c (hunk 1 above)
Comment (dgarske) [inline src/internal.c]: Good eye! Just pushed a fix. There was a reason for the change, due to set but not read error, but neglected to set ret on AllocDer. Thanks!

**Comment by toddouska** [comment]
Thanks.  As a future enhancement let's talk next week about putting the DerBuffers into a structure that can be released itself during handshake resource free.  Before the buffer was a pointer and size, now it's 2 pointers and 3 sizes (and there are 3 total).  We could save 88 bytes per connection or so on 64bit, or nearly 10% of post handshake resources.

**Pull request merged**